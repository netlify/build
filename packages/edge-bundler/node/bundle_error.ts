import type { ExecaError } from 'execa'

interface BundleErrorOptions {
  format: string
}

const getCustomErrorInfo = (options?: BundleErrorOptions) => ({
  location: {
    format: options?.format,
    runtime: 'deno',
  },
  type: 'functionsBundling',
})

class BundleError extends Error {
  customErrorInfo: ReturnType<typeof getCustomErrorInfo>

  constructor(originalError: Error, options?: BundleErrorOptions) {
    super(originalError.message)

    this.customErrorInfo = getCustomErrorInfo(options)
    this.name = 'BundleError'
    this.stack = originalError.stack

    // https://github.com/microsoft/TypeScript-wiki/blob/8a66ecaf77118de456f7cd9c56848a40fe29b9b4/Breaking-Changes.md#implicit-any-error-raised-for-un-annotated-callback-arguments-with-no-matching-overload-arguments
    Object.setPrototypeOf(this, BundleError.prototype)
  }
}

/**
 * BundleErrors are treated as user-error, so Netlify Team is not alerted about them.
 */
const wrapBundleError = (input: unknown, options?: BundleErrorOptions) => {
  if (input instanceof Error) {
    if (input.message.includes("The module's source code could not be parsed")) {
      input.message = (input as ExecaError).stderr
    }

    return new BundleError(input, options)
  }

  return input
}

export { BundleError, wrapBundleError }
