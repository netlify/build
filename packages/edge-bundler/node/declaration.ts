import { FunctionConfig, HTTPMethod, Path } from './config.js'
import { FeatureFlags } from './feature_flags.js'

interface BaseDeclaration {
  cache?: string
  function: string
  method?: HTTPMethod | HTTPMethod[]
  // todo: remove these two after a while and only support in-source config for non-route related configs
  name?: string
  generator?: string
}

type DeclarationWithPath = BaseDeclaration & {
  path: Path
  excludedPath?: Path | Path[]
}

type DeclarationWithPattern = BaseDeclaration & {
  pattern: string
  excludedPattern?: string | string[]
}

export type Declaration = DeclarationWithPath | DeclarationWithPattern

export const mergeDeclarations = (
  tomlDeclarations: Declaration[],
  userFunctionsConfig: Record<string, FunctionConfig>,
  internalFunctionsConfig: Record<string, FunctionConfig>,
  deployConfigDeclarations: Declaration[],
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  _featureFlags: FeatureFlags = {},
) => {
  const functionsVisited: Set<string> = new Set()

  const declarations: Declaration[] = [
    // INTEGRATIONS
    // 1. Declarations from the integrations deploy config
    ...getDeclarationsFromInput(deployConfigDeclarations, internalFunctionsConfig, functionsVisited),
    // 2. Declarations from the integrations ISC
    ...createDeclarationsFromFunctionConfigs(internalFunctionsConfig, functionsVisited),

    // USER
    // 3. Declarations from the users toml config
    ...getDeclarationsFromInput(tomlDeclarations, userFunctionsConfig, functionsVisited),
    // 4. Declarations from the users ISC
    ...createDeclarationsFromFunctionConfigs(userFunctionsConfig, functionsVisited),
  ]

  return declarations
}

const getDeclarationsFromInput = (
  inputDeclarations: Declaration[],
  functionConfigs: Record<string, FunctionConfig>,
  functionsVisited: Set<string>,
): Declaration[] => {
  const declarations: Declaration[] = []
  // For any declaration for which we also have a function configuration object,
  // we replace the path because that object takes precedence.
  for (const declaration of inputDeclarations) {
    const config = functionConfigs[declaration.function]

    if (!config) {
      // If no config is found, add the declaration as is.
      declarations.push(declaration)
    } else if (config.path?.length) {
      // If we have a path specified as either a string or non-empty array,
      // create a declaration for each path.
      const paths = Array.isArray(config.path) ? config.path : [config.path]

      paths.forEach((path) => {
        declarations.push({ ...declaration, cache: config.cache, path })
      })
    } else {
      // With an in-source config without a path, add the config to the declaration.
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { path, excludedPath, ...rest } = config

      declarations.push({ ...declaration, ...rest })
    }

    functionsVisited.add(declaration.function)
  }

  return declarations
}

const createDeclarationsFromFunctionConfigs = (
  functionConfigs: Record<string, FunctionConfig>,
  functionsVisited: Set<string>,
): Declaration[] => {
  const declarations: Declaration[] = []

  for (const name in functionConfigs) {
    const { cache, excludedPath, path, method } = functionConfigs[name]

    // If we have a path specified, create a declaration for each path.
    if (!functionsVisited.has(name) && path) {
      const paths = Array.isArray(path) ? path : [path]

      paths.forEach((singlePath) => {
        const declaration: Declaration = { function: name, path: singlePath }
        if (cache) {
          declaration.cache = cache
        }
        if (method) {
          declaration.method = method
        }
        if (excludedPath) {
          declaration.excludedPath = excludedPath
        }
        declarations.push(declaration)
      })
    }
  }

  return declarations
}

/**
 * Normalizes a regular expression, ensuring it has a leading `^` and trailing
 * `$` characters.
 */
export const normalizePattern = (pattern: string) => {
  let enclosedPattern = pattern

  if (!pattern.startsWith('^')) {
    enclosedPattern = `^${enclosedPattern}`
  }

  if (!pattern.endsWith('$')) {
    enclosedPattern = `${enclosedPattern}$`
  }

  const regexp = new RegExp(enclosedPattern)

  // Strip leading and forward slashes.
  return regexp.toString().slice(1, -1)
}
