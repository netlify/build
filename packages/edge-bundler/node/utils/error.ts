export const isNodeError = (error: any): error is NodeJS.ErrnoException => error instanceof Error

export const isFileNotFoundError = (error: any) => isNodeError(error) && error.code === 'ENOENT'
