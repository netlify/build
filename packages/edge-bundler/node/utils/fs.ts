import path from 'path'

/**
 * Returns all the directories obtained by traversing `inner` and its parents
 * all the way to `outer`, inclusive.
 */
export const pathsBetween = (inner: string, outer: string, paths: string[] = []): string[] => {
  const parent = path.dirname(inner)

  if (inner === outer || inner === parent) {
    return [...paths, outer]
  }

  return [inner, ...pathsBetween(parent, outer)]
}
