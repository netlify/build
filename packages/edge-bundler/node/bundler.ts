import { promises as fs } from 'fs'
import { join } from 'path'

import commonPathPrefix from 'common-path-prefix'
import { v4 as uuidv4 } from 'uuid'

import { importMapSpecifier } from '../shared/consts.js'

import { DenoBridge, DenoOptions, OnAfterDownloadHook, OnBeforeDownloadHook } from './bridge.js'
import type { Bundle } from './bundle.js'
import { FunctionConfig, getFunctionConfig } from './config.js'
import { Declaration, mergeDeclarations } from './declaration.js'
import { load as loadDeployConfig } from './deploy_config.js'
import { EdgeFunction } from './edge_function.js'
import { FeatureFlags, getFlags } from './feature_flags.js'
import { findFunctions } from './finder.js'
import { bundle as bundleESZIP } from './formats/eszip.js'
import { ImportMap } from './import_map.js'
import { getLogger, LogFunction, Logger } from './logger.js'
import { writeManifest } from './manifest.js'
import { vendorNPMSpecifiers } from './npm_dependencies.js'
import { ensureLatestTypes } from './types.js'

export interface BundleOptions {
  basePath?: string
  bootstrapURL?: string
  cacheDirectory?: string
  configPath?: string
  debug?: boolean
  distImportMapPath?: string
  featureFlags?: FeatureFlags
  importMapPaths?: (string | undefined)[]
  internalSrcFolder?: string
  onAfterDownload?: OnAfterDownloadHook
  onBeforeDownload?: OnBeforeDownloadHook
  rootPath?: string
  systemLogger?: LogFunction
  userLogger?: LogFunction
  vendorDirectory?: string
}

export const bundle = async (
  sourceDirectories: string[],
  distDirectory: string,
  tomlDeclarations: Declaration[] = [],
  {
    basePath: inputBasePath,
    cacheDirectory,
    configPath,
    debug,
    distImportMapPath,
    featureFlags: inputFeatureFlags,
    importMapPaths = [],
    internalSrcFolder,
    onAfterDownload,
    onBeforeDownload,
    rootPath,
    userLogger,
    systemLogger,
    vendorDirectory,
  }: BundleOptions = {},
) => {
  const logger = getLogger(systemLogger, userLogger, debug)
  const featureFlags = getFlags(inputFeatureFlags)
  const options: DenoOptions = {
    debug,
    cacheDirectory,
    logger,
    onAfterDownload,
    onBeforeDownload,
  }

  if (cacheDirectory !== undefined) {
    options.denoDir = join(cacheDirectory, 'deno_dir')
  }

  const deno = new DenoBridge(options)
  const basePath = getBasePath(sourceDirectories, inputBasePath)

  await ensureLatestTypes(deno, logger)

  // The name of the bundle will be the hash of its contents, which we can't
  // compute until we run the bundle process. For now, we'll use a random ID
  // to create the bundle artifacts and rename them later.
  const buildID = uuidv4()

  // Loading any configuration options from the deploy configuration API, if it
  // exists.
  const deployConfig = await loadDeployConfig(configPath, logger)

  // Layers are marked as externals in the ESZIP, so that those specifiers are
  // not actually included in the bundle.
  const externals = deployConfig.layers.map((layer) => layer.name)

  const userSourceDirectories = sourceDirectories.filter((dir) => dir !== internalSrcFolder)

  const importMap = new ImportMap()

  await importMap.addFiles([deployConfig?.importMap, ...importMapPaths], logger)

  const userFunctions = userSourceDirectories.length === 0 ? [] : await findFunctions(userSourceDirectories)
  const internalFunctions = internalSrcFolder ? await findFunctions([internalSrcFolder]) : []
  const functions = [...internalFunctions, ...userFunctions]
  const vendor = await safelyVendorNPMSpecifiers({
    basePath,
    featureFlags,
    functions,
    importMap,
    logger,
    rootPath: rootPath ?? basePath,
    vendorDirectory,
  })

  if (vendor) {
    importMap.add(vendor.importMap)
  }

  const functionBundle = await bundleESZIP({
    basePath,
    buildID,
    debug,
    deno,
    distDirectory,
    externals,
    functions,
    featureFlags,
    importMap,
    vendorDirectory: vendor?.directory,
  })

  // The final file name of the bundles contains a SHA256 hash of the contents,
  // which we can only compute now that the files have been generated. So let's
  // rename the bundles to their permanent names.
  await createFinalBundles([functionBundle], distDirectory, buildID)

  // Retrieving a configuration object for each function.
  // Run `getFunctionConfig` in parallel as it is a non-trivial operation and spins up deno
  const internalConfigPromises = internalFunctions.map(
    async (func) => [func.name, await getFunctionConfig({ func, importMap, deno, log: logger })] as const,
  )
  const userConfigPromises = userFunctions.map(
    async (func) => [func.name, await getFunctionConfig({ func, importMap, deno, log: logger })] as const,
  )

  // Creating a hash of function names to configuration objects.
  const internalFunctionsWithConfig = Object.fromEntries(await Promise.all(internalConfigPromises))
  const userFunctionsWithConfig = Object.fromEntries(await Promise.all(userConfigPromises))

  // Creating a final declarations array by combining the TOML file with the
  // deploy configuration API and the in-source configuration.
  const declarations = mergeDeclarations(
    tomlDeclarations,
    userFunctionsWithConfig,
    internalFunctionsWithConfig,
    deployConfig.declarations,
    featureFlags,
  )

  const internalFunctionConfig = createFunctionConfig({
    internalFunctionsWithConfig,
    declarations,
  })

  const manifest = await writeManifest({
    bundles: [functionBundle],
    declarations,
    distDirectory,
    featureFlags,
    functions,
    userFunctionConfig: userFunctionsWithConfig,
    internalFunctionConfig,
    importMap: importMapSpecifier,
    layers: deployConfig.layers,
  })

  await vendor?.cleanup()

  if (distImportMapPath) {
    await importMap.writeToFile(distImportMapPath)
  }

  return { functions, manifest }
}

const createFinalBundles = async (bundles: Bundle[], distDirectory: string, buildID: string) => {
  const renamingOps = bundles.map(async ({ extension, hash }) => {
    const tempBundlePath = join(distDirectory, `${buildID}${extension}`)
    const finalBundlePath = join(distDirectory, `${hash}${extension}`)

    await fs.rename(tempBundlePath, finalBundlePath)
  })

  await Promise.all(renamingOps)
}

const getBasePath = (sourceDirectories: string[], inputBasePath?: string) => {
  // If there's a specific base path supplied, that takes precedence.
  if (inputBasePath !== undefined) {
    return inputBasePath
  }

  // `common-path-prefix` returns an empty string when called with a single
  // path, so we check for that case and return the path itself instead.
  if (sourceDirectories.length === 1) {
    return sourceDirectories[0]
  }

  return commonPathPrefix(sourceDirectories)
}

interface MergeWithDeclarationConfigOptions {
  functionName: string
  config: FunctionConfig
  declarations: Declaration[]
}

// We used to allow the `name` and `generator` fields to be defined at the
// declaration level. We want these properties to live at the function level
// in their config object, so we translate that for backwards-compatibility.
const mergeWithDeclarationConfig = ({ functionName, config, declarations }: MergeWithDeclarationConfigOptions) => {
  const declaration = declarations?.find((decl) => decl.function === functionName)

  return {
    ...config,
    name: declaration?.name || config.name,
    generator: declaration?.generator || config.generator,
  }
}

const addGeneratorFallback = (config: FunctionConfig) => ({
  ...config,
  generator: config.generator || 'internalFunc',
})

interface CreateFunctionConfigOptions {
  internalFunctionsWithConfig: Record<string, FunctionConfig>
  declarations: Declaration[]
}

const createFunctionConfig = ({ internalFunctionsWithConfig, declarations }: CreateFunctionConfigOptions) =>
  Object.entries(internalFunctionsWithConfig).reduce((acc, [functionName, config]) => {
    const mergedConfigFields = mergeWithDeclarationConfig({ functionName, config, declarations })

    return {
      ...acc,
      [functionName]: addGeneratorFallback(mergedConfigFields),
    }
  }, {} as Record<string, FunctionConfig>)

interface VendorNPMOptions {
  basePath: string
  featureFlags: FeatureFlags
  functions: EdgeFunction[]
  importMap: ImportMap
  logger: Logger
  rootPath: string
  vendorDirectory: string | undefined
}

const safelyVendorNPMSpecifiers = async ({
  basePath,
  functions,
  importMap,
  logger,
  rootPath,
  vendorDirectory,
}: VendorNPMOptions) => {
  try {
    return await vendorNPMSpecifiers({
      basePath,
      directory: vendorDirectory,
      functions: functions.map(({ path }) => path),
      importMap,
      logger,
      environment: 'production',
      rootPath,
    })
  } catch (error) {
    logger.system(error)
  }
}
