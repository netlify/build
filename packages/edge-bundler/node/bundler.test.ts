import { Buffer } from 'buffer'
import { access, readdir, readFile, rm, writeFile } from 'fs/promises'
import { join, resolve } from 'path'
import process from 'process'
import { pathToFileURL } from 'url'

import tmp from 'tmp-promise'
import { test, expect, vi } from 'vitest'

import { importMapSpecifier } from '../shared/consts.js'
import { runESZIP, useFixture } from '../test/util.js'

import { BundleError } from './bundle_error.js'
import { bundle, BundleOptions } from './bundler.js'
import { Declaration } from './declaration.js'
import { isFileNotFoundError } from './utils/error.js'
import { validateManifest } from './validation/manifest/index.js'

test('Produces an ESZIP bundle', async () => {
  const { basePath, cleanup, distPath } = await useFixture('with_import_maps')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const userDirectory = join(basePath, 'user-functions')
  const internalDirectory = join(basePath, 'functions')
  const result = await bundle([userDirectory, internalDirectory], distPath, declarations, {
    basePath,
    configPath: join(internalDirectory, 'config.json'),
    importMapPaths: [join(userDirectory, 'import_map.json')],
  })
  const generatedFiles = await readdir(distPath)

  expect(result.functions.length).toBe(3)
  expect(generatedFiles.length).toBe(2)

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)
  expect(() => validateManifest(manifest)).not.toThrowError()
  const { bundles, import_map: importMapURL } = manifest

  expect(bundles.length).toBe(1)
  expect(bundles[0].format).toBe('eszip2')
  expect(generatedFiles.includes(bundles[0].asset)).toBe(true)

  expect(importMapURL).toBe(importMapSpecifier)

  const bundlePath = join(distPath, bundles[0].asset)

  const { func1, func2, func3 } = await runESZIP(bundlePath)

  expect(func1).toBe('HELLO, JANE DOE!')
  expect(func2).toBe('Jane Doe')
  expect(func3).toBe('hello, netlify!')

  await cleanup()
})

test('Uses the vendored eszip module instead of fetching it from deno.land', async () => {
  const { basePath, cleanup, distPath } = await useFixture('with_import_maps')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const sourceDirectory = join(basePath, 'functions')
  const result = await bundle([sourceDirectory], distPath, declarations, {
    basePath,
    configPath: join(sourceDirectory, 'config.json'),
  })
  const generatedFiles = await readdir(distPath)

  expect(result.functions.length).toBe(1)
  expect(generatedFiles.length).toBe(2)

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)
  const { bundles } = manifest

  expect(bundles.length).toBe(1)
  expect(bundles[0].format).toBe('eszip2')
  expect(generatedFiles.includes(bundles[0].asset)).toBe(true)

  await cleanup()
})

test('Adds a custom error property to user errors during bundling', async () => {
  process.env.NO_COLOR = 'true'
  expect.assertions(3)

  const { basePath, cleanup, distPath } = await useFixture('invalid_functions')
  const sourceDirectory = join(basePath, 'functions')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]

  try {
    await bundle([sourceDirectory], distPath, declarations, { basePath })
  } catch (error) {
    expect(error).toBeInstanceOf(BundleError)
    const [messageBeforeStack] = (error as BundleError).message.split('at <anonymous> (file://')
    expect(messageBeforeStack).toMatchInlineSnapshot(`
      "error: Uncaught (in promise) Error: The module's source code could not be parsed: Unexpected eof at file:///root/functions/func1.ts:1:27

        export default async () => 
                                  ~
            const ret = new Error(getStringFromWasm0(arg0, arg1));
                        ^
          "
    `)
    expect((error as BundleError).customErrorInfo).toEqual({
      location: {
        format: 'eszip',
        runtime: 'deno',
      },
      type: 'functionsBundling',
    })
  } finally {
    await cleanup()
  }
})

test('Prints a nice error message when user tries importing an npm module', async () => {
  expect.assertions(2)

  const { basePath, cleanup, distPath } = await useFixture('imports_npm_module_scheme')
  const sourceDirectory = join(basePath, 'functions')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]

  try {
    await bundle([sourceDirectory], distPath, declarations, {
      basePath,
    })
  } catch (error) {
    expect(error).toBeInstanceOf(BundleError)
    expect((error as BundleError).message).toEqual(
      `There was an error when loading the 'p-retry' npm module. Support for npm modules in edge functions is an experimental feature. Refer to https://ntl.fyi/edge-functions-npm for more information.`,
    )
  } finally {
    await cleanup()
  }
})

test('Does not add a custom error property to system errors during bundling', async () => {
  expect.assertions(1)

  try {
    // @ts-expect-error Sending bad input to `bundle` to force a system error.
    await bundle([123, 321], '/some/directory', declarations)
  } catch (error) {
    expect(error).not.toBeInstanceOf(BundleError)
  }
})

test('Uses the cache directory as the `DENO_DIR` value', async () => {
  expect.assertions(3)

  const { basePath, cleanup, distPath } = await useFixture('with_import_maps')
  const sourceDirectory = join(basePath, 'functions')
  const cacheDir = await tmp.dir()
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const options: BundleOptions = {
    basePath,
    cacheDirectory: cacheDir.path,
    configPath: join(sourceDirectory, 'config.json'),
  }

  const result = await bundle([sourceDirectory], distPath, declarations, options)
  const outFiles = await readdir(distPath)

  expect(result.functions.length).toBe(1)
  expect(outFiles.length).toBe(2)

  const denoDir = await readdir(join(cacheDir.path, 'deno_dir'))

  expect(denoDir.includes('gen')).toBe(true)

  await cleanup()
})

test('Supports import maps with relative paths', async () => {
  const { basePath, cleanup, distPath } = await useFixture('with_import_maps')
  const sourceDirectory = join(basePath, 'functions')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const result = await bundle([sourceDirectory], distPath, declarations, {
    basePath,
    configPath: join(sourceDirectory, 'config.json'),
  })
  const generatedFiles = await readdir(distPath)

  expect(result.functions.length).toBe(1)
  expect(generatedFiles.length).toBe(2)

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)
  const { bundles } = manifest

  expect(bundles.length).toBe(1)
  expect(bundles[0].format).toBe('eszip2')
  expect(generatedFiles.includes(bundles[0].asset)).toBe(true)

  await cleanup()
})

test('Ignores any user-defined `deno.json` files', async () => {
  const { basePath, cleanup, distPath } = await useFixture('with_import_maps')
  const sourceDirectory = join(basePath, 'functions')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]

  // Creating an import map file that rewires the URL of the Deno registry to
  // an invalid location.
  const importMapFile = await tmp.file()
  const importMap = {
    imports: {
      'https://deno.land/': 'https://black.hole/',
    },
  }

  await writeFile(importMapFile.path, JSON.stringify(importMap))

  // Deno configuration files need to be in the current working directory.
  // There's not a great way for us to set the working directory of the `deno`
  // process that we'll run, so our best bet is to write the file to whatever
  // is the current working directory now and then clean it up.
  const denoConfigPath = join(process.cwd(), 'deno.json')
  const denoConfig = {
    importMap: importMapFile.path,
  }

  try {
    await access(denoConfigPath)

    throw new Error(
      `The file at '${denoConfigPath} would be overwritten by this test. Please move the file to a different location and try again.'`,
    )
  } catch (error) {
    if (!isFileNotFoundError(error)) {
      throw error
    }
  }

  await writeFile(denoConfigPath, JSON.stringify(denoConfig))

  expect(() =>
    bundle([sourceDirectory], distPath, declarations, {
      basePath,
      configPath: join(sourceDirectory, 'config.json'),
    }),
  ).not.toThrow()

  await cleanup()
  await rm(denoConfigPath, { force: true, recursive: true, maxRetries: 10 })
  await rm(importMapFile.path, { force: true, recursive: true, maxRetries: 10 })
})

test('Processes a function that imports a custom layer', async () => {
  const { basePath, cleanup, distPath } = await useFixture('with_layers')
  const sourceDirectory = join(basePath, 'functions')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const layer = { name: 'https://edge-function-layer-template.netlify.app/mod.ts', flag: 'edge-functions-layer-test' }
  const result = await bundle([sourceDirectory], distPath, declarations, {
    basePath,
    configPath: join(sourceDirectory, 'config.json'),
  })
  const generatedFiles = await readdir(distPath)

  expect(result.functions.length).toBe(1)
  expect(generatedFiles.length).toBe(2)

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)
  const { bundles, layers } = manifest

  expect(bundles.length).toBe(1)
  expect(bundles[0].format).toBe('eszip2')
  expect(generatedFiles.includes(bundles[0].asset)).toBe(true)

  expect(layers).toEqual([layer])

  await cleanup()
})

test('Loads declarations and import maps from the deploy configuration and in-source config', async () => {
  const { basePath, cleanup, distPath } = await useFixture('with_deploy_config')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const directories = [join(basePath, 'netlify', 'edge-functions'), join(basePath, '.netlify', 'edge-functions')]
  const result = await bundle(directories, distPath, declarations, {
    basePath,
    configPath: join(basePath, '.netlify', 'edge-functions', 'manifest.json'),
    internalSrcFolder: directories[1],
  })
  const generatedFiles = await readdir(distPath)

  expect(result.functions.length).toBe(3)
  expect(generatedFiles.length).toBe(2)

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)
  const { bundles, function_config: functionConfig, routes } = manifest
  expect(bundles.length).toBe(1)
  expect(bundles[0].format).toBe('eszip2')
  expect(generatedFiles.includes(bundles[0].asset)).toBe(true)

  // respects excludedPath from deploy config
  expect(routes[0].excluded_patterns).toEqual(['^/func2/skip/?$'])

  expect(functionConfig.func2).toEqual({
    name: 'Function two',
    generator: '@netlify/fake-plugin@1.0.0',
  })

  // respects in-source config
  expect(functionConfig.func3).toEqual({
    name: 'in-config-function',
    on_error: 'bypass',
    generator: 'internalFunc',
  })

  await cleanup()
})

test("Ignores entries in `importMapPaths` that don't point to an existing import map file", async () => {
  const systemLogger = vi.fn()
  const { basePath, cleanup, distPath } = await useFixture('with_import_maps')
  const sourceDirectory = join(basePath, 'user-functions')

  // Creating import map file
  const importMap = await tmp.file()
  const importMapContents = {
    imports: {
      helper: pathToFileURL(join(basePath, 'helper.ts')).toString(),
    },
    scopes: {
      [pathToFileURL(join(sourceDirectory, 'func3/func3.ts')).toString()]: {
        helper: pathToFileURL(join(basePath, 'helper2.ts')).toString(),
      },
    },
  }

  await writeFile(importMap.path, JSON.stringify(importMapContents))

  const nonExistingImportMapPath = join(distPath, 'some-file-that-does-not-exist.json')
  const result = await bundle(
    [sourceDirectory],
    distPath,
    [
      {
        function: 'func1',
        path: '/func1',
      },
    ],
    {
      basePath,
      importMapPaths: [nonExistingImportMapPath, importMap.path],
      systemLogger,
    },
  )
  const generatedFiles = await readdir(distPath)

  expect(result.functions.length).toBe(2)
  expect(generatedFiles.length).toBe(2)
  expect(systemLogger).toHaveBeenCalledWith(`Did not find an import map file at '${nonExistingImportMapPath}'.`)

  await cleanup()
  await importMap.cleanup()
})

test('Handles imports with the `node:` prefix', async () => {
  const { basePath, cleanup, distPath } = await useFixture('imports_node_specifier')
  const userDirectory = join(basePath, 'netlify', 'edge-functions')
  const result = await bundle([userDirectory], distPath, [], {
    basePath,
    importMapPaths: [join(userDirectory, 'import_map.json')],
  })
  const generatedFiles = await readdir(distPath)

  expect(result.functions.length).toBe(1)
  expect(generatedFiles.length).toBe(2)

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)

  expect(() => validateManifest(manifest)).not.toThrowError()

  const { bundles, import_map: importMapURL, routes } = manifest

  expect(bundles.length).toBe(1)
  expect(bundles[0].format).toBe('eszip2')
  expect(generatedFiles.includes(bundles[0].asset)).toBe(true)
  expect(importMapURL).toBe(importMapSpecifier)
  expect(routes.length).toBe(1)
  expect(routes[0].function).toBe('func1')
  expect(routes[0].pattern).toBe('^/func1/?$')

  const bundlePath = join(distPath, bundles[0].asset)

  const { func1 } = await runESZIP(bundlePath)

  expect(func1).toBe('ok')

  await cleanup()
})

test('Loads npm modules from bare specifiers', async () => {
  const systemLogger = vi.fn()
  const { basePath, cleanup, distPath } = await useFixture('imports_npm_module')
  const sourceDirectory = join(basePath, 'functions')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const vendorDirectory = await tmp.dir()

  await bundle([sourceDirectory], distPath, declarations, {
    basePath,
    importMapPaths: [join(basePath, 'import_map.json')],
    vendorDirectory: vendorDirectory.path,
    systemLogger,
  })

  expect(
    systemLogger.mock.calls.find((call) => call[0] === 'Could not track dependencies in edge function:'),
  ).toBeUndefined()

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)
  const bundlePath = join(distPath, manifest.bundles[0].asset)
  const { func1 } = await runESZIP(bundlePath, vendorDirectory.path)

  expect(func1).toBe(
    `<parent-1><child-1>JavaScript</child-1></parent-1>, <parent-2><child-2><grandchild-1>APIs<cwd>${process.cwd()}</cwd></grandchild-1></child-2></parent-2>, <parent-3><child-2><grandchild-1>Markup<cwd>${process.cwd()}</cwd></grandchild-1></child-2></parent-3>`,
  )

  await cleanup()
  await rm(vendorDirectory.path, { force: true, recursive: true })
})

test('Loads npm modules in a monorepo setup', async () => {
  const systemLogger = vi.fn()
  const { basePath: rootPath, cleanup, distPath } = await useFixture('monorepo_npm_module')
  const basePath = join(rootPath, 'packages', 'frontend')
  const sourceDirectory = join(basePath, 'functions')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const vendorDirectory = await tmp.dir()

  await bundle([sourceDirectory], distPath, declarations, {
    basePath,
    importMapPaths: [join(basePath, 'import_map.json')],
    rootPath,
    vendorDirectory: vendorDirectory.path,
    systemLogger,
  })

  expect(
    systemLogger.mock.calls.find((call) => call[0] === 'Could not track dependencies in edge function:'),
  ).toBeUndefined()

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)
  const bundlePath = join(distPath, manifest.bundles[0].asset)
  const { func1 } = await runESZIP(bundlePath, vendorDirectory.path)

  expect(func1).toBe(
    `<parent-1><child-1>JavaScript</child-1></parent-1>, <parent-2><child-2><grandchild-1>APIs<cwd>${process.cwd()}</cwd></grandchild-1></child-2></parent-2>, <parent-3><child-2><grandchild-1>Markup<cwd>${process.cwd()}</cwd></grandchild-1></child-2></parent-3>`,
  )

  await cleanup()
  await rm(vendorDirectory.path, { force: true, recursive: true })
})

test('Loads JSON modules', async () => {
  const { basePath, cleanup, distPath } = await useFixture('imports_json')
  const sourceDirectory = join(basePath, 'functions')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const vendorDirectory = await tmp.dir()

  await bundle([sourceDirectory], distPath, declarations, {
    basePath,
    vendorDirectory: vendorDirectory.path,
  })

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)
  const bundlePath = join(distPath, manifest.bundles[0].asset)
  const { func1 } = await runESZIP(bundlePath, vendorDirectory.path)

  expect(func1).toBe(`{"foo":"bar"}`)

  await cleanup()
  await rm(vendorDirectory.path, { force: true, recursive: true })
})

test('Supports TSX and process.env', async () => {
  const { basePath, cleanup, distPath } = await useFixture('tsx')
  const sourceDirectory = join(basePath, 'functions')
  const declarations: Declaration[] = [
    {
      function: 'func1',
      path: '/func1',
    },
  ]
  const vendorDirectory = await tmp.dir()

  await bundle([sourceDirectory], distPath, declarations, {
    basePath,
    vendorDirectory: vendorDirectory.path,
  })

  const manifestFile = await readFile(resolve(distPath, 'manifest.json'), 'utf8')
  const manifest = JSON.parse(manifestFile)
  const bundlePath = join(distPath, manifest.bundles[0].asset)
  process.env.FOO = 'bar'
  const { func1 } = await runESZIP(bundlePath, vendorDirectory.path)

  expect(Buffer.from(func1, 'base64').toString()).toBe(
    `hippedy hoppedy, createElement is now a production property. Here, take this env var: FOO=bar`,
  )

  await cleanup()
  await rm(vendorDirectory.path, { force: true, recursive: true })
  delete process.env.FOO
})
