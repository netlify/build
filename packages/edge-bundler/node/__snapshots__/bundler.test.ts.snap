// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Adds a custom error property to user errors during bundling 1`] = `
"error: Uncaught (in promise) Error: The module's source code could not be parsed: Unexpected eof at file:///root/functions/func1.ts:1:27

  export default async () => 
                            ~
      const ret = new Error(getStringFromWasm0(arg0, arg1));
                  ^
    at <anonymous> (file://build/packages/edge-bundler/deno/vendor/deno.land/x/eszip@v0.55.2/eszip_wasm.generated.js:443:19)"
`;
